### PR TITLE
moduleinst is taken out into a separated cell & call by index and identifier supported

### PR DESCRIPTION
**Pull request by QianyangPeng**
The moduleinst is taken out into a separated cell.

Now we have the better module instantiation and the indices are allocated incrementally from zero.

The identifiers, indices and addresses are better supported. Now functions should be defined with identifiers and could be referenced by either identifiers or indices.

### CODE REVIEW

File: wasm.md [hunk 1]
```diff
@@ -125,6 +128,20 @@ Function `#unsigned` is called on integers to allow programs to use negative num
...
+To resolve these `identifiers` into concrete `indices`, some grammar production are indexed by an identifier context `I` as a synthesized attribute that records the declared identifiers in each index space. We call this operation `ICov`.
+
+```k
+    syntax Int ::= #ContextLookup  ( Map , TextFormatIdx ) [function]
+                 | #asInt ( KItem ) [function]
+ // -----------------------------------------------------------------
+    rule #asInt(I:Int) => I
+    rule #ContextLookup(IDS:Map, I:Int) => I
+    rule #ContextLookup(IDS:Map, ID:Identifier) => #asInt( IDS [ ID ] )
+```
```
Comment (ehildenb) [inline wasm.md]: We can do this simpler using K type-casting:

```
    syntax Int ::= #ContextLookup ( Map , TextFormatIdx ) [function] 
 // ---------------------------------------------------------------- 
    rule #ContextLookup(IDS:Map, I:Int)         => I 
    rule #ContextLookup(IDS:Map, ID:Identifier) => {IDS [ ID ]}:>Int 
```

File: wasm.md [hunk 2]
```diff
@@ -718,15 +738,16 @@ The allocation of a new `tableinst`. Currently at most one table may be defined
...
-         <tabAddrs>    .Map => (0 |-> NEXT)  </tabAddrs>
-         <nextTabAddr> NEXT => NEXT +Int 1 </nextTabAddr>
+         <nextTabIdx> NEXTIDX => NEXTIDX +Int 1 </nextTabIdx>
+         <tabIndices> .Map => (NEXTIDX |-> NEXTADDR) </tabIndices>
+         <nextTabAddr> NEXTADDR => NEXTADDR +Int 1 </nextTabAddr>
          <tabs>
            ( .Bag
           => <tabInst>
-               <tAddr>   NEXT </tAddr>
+               <tAddr>   NEXTADDR </tAddr>
```
Comment (ehildenb) [inline wasm.md]: Formatting, line up the closing cell tags.

File: wasm.md [hunk 3]
```diff
@@ -759,15 +780,16 @@ Currently, only one memory may be accessible to a module, and thus the `<memAddr
...
-         <memAddrs>    .Map => (0 |-> NEXT)  </memAddrs>
-         <nextMemAddr> NEXT => NEXT +Int 1 </nextMemAddr>
+         <nextMemIdx> NEXTIDX => NEXTIDX +Int 1 </nextMemIdx>
+         <memIndices> .Map => (NEXTIDX |-> NEXTADDR) </memIndices>
+         <nextMemAddr> NEXTADDR => NEXTADDR +Int 1 </nextMemAddr>
          <mems>
            ( .Bag
           => <memInst>
-               <mAddr>   NEXT </mAddr>
+               <mAddr>   NEXTADDR </mAddr>
```
Comment (ehildenb) [inline wasm.md]: formatting

File: wasm.md (hunk 1 above)
Comment (QianyangPeng) [reply wasm.md]: fixed

File: wasm.md (hunk 3 above)
Comment (QianyangPeng) [reply wasm.md]: fixed

File: wasm.md (hunk 2 above)
Comment (QianyangPeng) [reply wasm.md]: fixed

File: test.md [hunk 4]
```diff
@@ -131,12 +132,13 @@ This asserts related operation about tables.
...
                        | "#assertEmptyTableAux" Int Int MaxBound String
- // --------------------------------------------------------------------
+ // -------------------------------------------------------------------
     rule <k> #assertEmptyTable SIZE MAX MSG => #assertEmptyTableAux (NEXT -Int 1) SIZE MAX MSG ... </k>
          <nextTabAddr> NEXT </nextTabAddr>
 
     rule <k> #assertEmptyTableAux ADDR SIZE MAX _ => .  ... </k>
-         <tabAddrs> (0 |-> ADDR) => .Map </tabAddrs>
+         <nextTabIdx> NEXT => NEXT -Int 1 </nextTabIdx>
+         <tabIndices> ... ( 0 |-> ADDR ) => .Map ... </tabIndices>
```
Comment (ehildenb) [inline test.md]: The `<tabIndinces>` should only ever have a single element for now right? Do we need the `...`?

File: test.md [hunk 5]
```diff
@@ -164,7 +166,8 @@ This checks that the last allocated memory has the given size and max value.
          <nextMemAddr> NEXT </nextMemAddr>
 
     rule <k> #assertEmptyMemoryAux ADDR SIZE MAX _ => .  ... </k>
-         <memAddrs> (0 |-> ADDR) => .Map </memAddrs>
+         <nextMemIdx> NEXT => NEXT -Int 1 </nextMemIdx>
+         <memIndices> ... ( 0 |-> ADDR ) => .Map ... </memIndices>
```
Comment (ehildenb) [inline test.md]: Same here.

File: wasm.md [hunk 6]
```diff
@@ -125,6 +128,19 @@ Function `#unsigned` is called on integers to allow programs to use negative num
...
+The text format allows the use of symbolic `identifiers` in place of `indices`.
+To resolve these `identifiers` into concrete `indices`, some grammar production are indexed by an identifier context `I` as a synthesized attribute that records the declared identifiers in each index space. We call this operation `ICov`.
+
+```k
+    syntax Int ::= #ContextLookup  ( Map , TextFormatIdx ) [function]
+                 | #asInt ( KItem ) [function]
+ // -----------------------------------------------------------------
+    rule #ContextLookup(IDS:Map, I:Int) => I
+    rule #ContextLookup(IDS:Map, ID:Identifier) => {IDS [ ID ]}:>Int 
+```
```
Comment (ehildenb) [inline wasm.md]: We don't need `#asInt` declared anymore. Also check your formatting for `#ContextLookup`.

File: test.md (hunk 4 above)
Comment (QianyangPeng) [reply test.md]: No. I fixed it

File: test.md (hunk 5 above)
Comment (QianyangPeng) [reply test.md]: fixed

File: wasm.md (hunk 6 above)
Comment (QianyangPeng) [reply wasm.md]: removed #asInt and fixed the format

**Pull request merged**